presto-plan-checker-providers/src/main/java/com/facebook/presto/plancheckerproviders/nativechecker/NativePlanChecker.java

```diff
@@ -0,0 +1,182 @@
+
+import static com.facebook.presto.spi.StandardErrorCode.NOT_FOUND;
+import static com.facebook.presto.spi.StandardErrorCode.QUERY_REJECTED;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Uses the native sidecar to check verify a plan can be run on a native worker.
+ */
+public final class NativePlanChecker
+        implements PlanChecker
+{
+    private static final Logger LOG = Logger.get(NativePlanChecker.class);
+    private static final String PRESTO_QUERY_ID = "X-Presto-Query-Id";
+    private static final String PRESTO_TIME_ZONE = "X-Presto-Time-Zone";
+    private static final String PRESTO_SYSTEM_PROPERTY = "X-Presto-System-Property";
+    private static final String PRESTO_CATALOG_PROPERTY = "X-Presto-Catalog-Property";
+    private static final MediaType JSON_CONTENT_TYPE = MediaType.parse("application/json; charset=utf-8");
+    public static final String PLAN_CONVERSION_ENDPOINT = "/v1/velox/plan";
+
+    private final NodeManager nodeManager;
+    private final JsonCodec<SimplePlanFragment> planFragmentJsonCodec;
+    private final NativePlanCheckerConfig config;
+    private final OkHttpClient httpClient;
+
+    public NativePlanChecker(NodeManager nodeManager, JsonCodec<SimplePlanFragment> planFragmentJsonCodec, NativePlanCheckerConfig config)
+    {
+        this.nodeManager = requireNonNull(nodeManager, "nodeManager is null");
+        this.planFragmentJsonCodec = requireNonNull(planFragmentJsonCodec, "planFragmentJsonCodec is null");
+        this.config = requireNonNull(config, "config is null");
+        this.httpClient = new OkHttpClient.Builder().build();
+    }
+
+    private URL getPlanValidateUrl()
+    {
+        Set<Node> nodes = nodeManager.getAllNodes();
+        Node sidecarNode = nodes.stream().filter(Node::isCoordinatorSidecar).findFirst()
+                .orElseThrow(() -> new PrestoException(NOT_FOUND, "could not find native sidecar node"));
+        try {
+            // endpoint to retrieve session properties from native worker
+            return new URL(sidecarNode.getHttpUri().toString() + PLAN_CONVERSION_ENDPOINT);
+        }
+        catch (MalformedURLException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Override
+    public void validate(PlanNode planNode, WarningCollector warningCollector)
+    {
+        // NO-OP, only validating fragments
+    }
+
+    @Override
+    public void validateFragment(SimplePlanFragment planFragment, WarningCollector warningCollector)
+    {
+        try {
+            if (!planFragment.getPartitioning().isCoordinatorOnly() && !isInternalSystemConnector(planFragment.getRoot())) {
+                runValidation(planFragment);
+            }
+            else if (LOG.isDebugEnabled()) {
+                LOG.debug("Skipping Native Plan Validation for plan fragment id: %s", planFragment.getId());
+            }
+        }
+        catch (final PrestoException e) {
+            if (config.isQueryFailOnError()) {
```

Comment:
My original thought was that if a particular query was not being validated correctly and causing a false negative, then FailOnError could be disabled and there would still be logs of ones that failed. That is pretty iffy though, so I'm totally fine removing this.

Reply:
I do want to confirm the expected behavior when a native sidecar node is not present. Currently, it will throw an IllegalStateException, if QueryFailOnError is enabled.

Reply:
> My original thought was that if a particular query was not being validated correctly and causing a false negative, then FailOnError could be disabled and there would still be logs of ones that failed. That is pretty iffy though, so I'm totally fine removing this.

Yeah, i think it adds complexity without being that useful. Better to let it fail or disable the feature.  

> I do want to confirm the expected behavior when a native sidecar node is not present. Currently, it will throw an IllegalStateException, if QueryFailOnError is enabled.

What failure would occur if we don't catch anything here? Probably we should just let that propagate without changing the error.

Reply:
> What failure would occur if we don't catch anything here? Probably we should just let that propagate without changing the error.

If it can't find the sidecar, it will throw `new PrestoException(NOT_FOUND, "could not find native sidecar node")`, I just want to confirm that is the right behavior?

Reply:
That sounds good to me.

Reply:
removed now